syft/pkg/catalog.go

```diff
@@ -41,43 +39,133 @@ func (c *Catalog) PackageCount() int {
 
 // Package returns the package with the given ID.
 func (c *Catalog) Package(id ID) *Package {
-	return c.byID[id]
+	v, exists := c.byID[id]
+	if !exists {
+		return nil
+	}
+	return v
+}
+
+// PackagesByPath returns all packages that were discovered from the given path.
+func (c *Catalog) PackagesByPath(path string) []*Package {
+	return c.Packages(c.idsByPath[path])
 }
 
-// PackagesByFile returns all packages that were discovered from the given source file reference.
-func (c *Catalog) PackagesByFile(location source.Location) []*Package {
-	return c.byFile[location]
+// Packages returns all packages for the given ID.
+func (c *Catalog) Packages(ids []ID) (result []*Package) {
+	for _, i := range ids {
+		p, exists := c.byID[i]
+		if exists {
+			result = append(result, p)
+		}
+	}
+	return result
 }
 
 // Add a package to the Catalog.
 func (c *Catalog) Add(p Package) {
-	if p.id != 0 {
-		log.Errorf("package already added to catalog: %s", p)
-		return
-	}
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	p.id = ID(nextPackageID)
-	nextPackageID++
+	_, exists := c.byID[p.ID]
+	if exists {
+		log.Errorf("package ID already exists in the catalog : id=%+v %+v", p.ID, p)
+		return
+	}
+
+	if p.ID == "" {
+		p.ID = newID()
+	}
 
 	// store by package ID
-	c.byID[p.id] = &p
+	c.byID[p.ID] = &p
 
 	// store by package type
-	_, ok := c.byType[p.Type]
-	if !ok {
-		c.byType[p.Type] = make([]*Package, 0)
+	c.idsByType[p.Type] = append(c.idsByType[p.Type], p.ID)
+
+	// store by file location paths
+	for _, l := range p.Locations {
+		if l.RealPath != "" {
+			c.idsByPath[l.RealPath] = append(c.idsByPath[l.RealPath], p.ID)
+		}
+		if l.VirtualPath != "" {
+			c.idsByPath[l.VirtualPath] = append(c.idsByPath[l.VirtualPath], p.ID)
+		}
+	}
+}
+
+func (c *Catalog) Remove(id ID) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	_, exists := c.byID[id]
+	if !exists {
+		log.Errorf("package ID does not exist in the catalog : id=%+v", id)
+		return
 	}
-	c.byType[p.Type] = append(c.byType[p.Type], &p)
 
-	// store by file references
-	for _, s := range p.Locations {
-		_, ok := c.byFile[s]
+	// Remove all index references to this package ID
+	for t, ids := range c.idsByType {
+		c.idsByType[t] = removeID(id, ids)
+		if len(c.idsByType[t]) == 0 {
+			delete(c.idsByType, t)
+		}
+	}
+
+	for p, ids := range c.idsByPath {
+		c.idsByPath[p] = removeID(id, ids)
+		if len(c.idsByPath[p]) == 0 {
+			delete(c.idsByPath, p)
+		}
+	}
+
+	// Remove package
+	delete(c.byID, id)
+}
+
+func (c *Catalog) Finalize() {
+	c.markPackageOwnership()
+}
+
+// markPackageOwnership find overlaps in file ownership with a file that defines another package. Specifically, a .Location.Path of
+// a package is found to be owned by another (from the owner's .Metadata.Files[]). This relationship is captured on the
+// child package.
+func (c *Catalog) markPackageOwnership() {
```

Comment:
This method is a bit on the long and nested side, it'd be nice to find ways to decompose this into more focused functions

Reply:
the "splits" I considered seemed contrived, so I instead added comments to help.

Reply:
The comments are a good start, because you've taken time to think through what each section of the method is doing. It's clear that different sections of this method are doing specific things. These comments can now provide a roadmap toward what makes sense to factor out. This will help others wrap their head around what's going on in this function more readily

Reply:
I guess what I'm saying is that I don't think this function is that long (35-ish lines including whitespace & comments).

I see where it __could__ get broken up into other functions, however in all cases you will be passing what could stay as local state around to other functions (`pkgParents`), which means a reader needs to understand what the separate function is doing relative to the already-prepared state from another function. In these cases I tend to prefer a slightly longer function (which is still within the linter rules) with healthy comments.

If there were multiple rules for what determines "ownership" I would probably change my mind.

Reply:
👍 We can move forward with this PR. We'll chat offline and follow up in another PR as necessary.